File: sloth.py
from sys import exit
from os import path
from scanner import Scanner
from token import Token

from typing import List, TextIO


class Sloth:
    had_error: bool = False

    @staticmethod
    def main(*args: List[str]):
        if len(args) > 1:
            print('Usage: sloth [script]')
            exit(64)
        elif len(args) == 1:
            Sloth.__run_file(args[0])
        else:
            Sloth.__run_prompt()

    @staticmethod
    def __run_file(rel_path: str):
        handler: TextIO = open(path.relpath(rel_path), 'r')
        file_str: str = handler.read()
        handler.close()
        Sloth.__run(file_str)
        if Sloth.had_error:
            exit(65)

    @staticmethod
    def __run_prompt():
        input_str: str = str(input())
        while True:
            print('> ', end='')
            Sloth.__run(input_str)
            Sloth.had_error = False

    @staticmethod
    def __run(source: str):
        scanner: Scanner = Scanner(source)
        tokens: List[Token] = scanner.scan_tokens()

        for token in tokens:
            print(token)

    @staticmethod
    def error(line: int, message: str):
        Sloth.__report(line, '', message)

    @staticmethod
    def __report(line: int, where: str, message: str):
        print('[line %s ] Error %s : %s' % line % where % message)
        Sloth.had_error = True


if __name__ == '__main__':
    Sloth.main()

File: sloth.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"strings"

	tokenizer "./tokenizer"
)

// Globals
var hadError bool = false

// Error function prints error to the Std Error
func Error(line int, message string) {
	report(line, "", message)
}

func report(line int, where string, message string) {
	fmt.Fprintf(os.Stderr, "[line %d] Error %s : %s", line, where, message)
	hadError = true
}

func run(source string) {
	t := tokenizer.NewToken(tokenizer.EOF, "", "", 23)
	fmt.Println(t.ToString())
}

func runFile(path string) {
	data, err := ioutil.ReadFile(path)
	if err != nil {
		panic(err)
	}
	run(string(data))
	if hadError {
		os.Exit(65)
	}
}

func runPrompt() {
	reader := bufio.NewReader(os.Stdin)

	for {
		fmt.Print("sloth> ")
		line, err := reader.ReadString('\n')
		if err != nil {
			panic(err)
		}
		if strings.TrimSpace(line) == "exit()" {
			os.Exit(0)
		}
		run(line)
		hadError = false
	}
}

func main() {
	var arguments []string = os.Args[1:]

	if argumentsLength := len(arguments); argumentsLength > 1 {
		fmt.Println("Usage: sloth [script]")
		os.Exit(64)
	} else if argumentsLength == 1 {
		runFile(arguments[0])
	} else {
		runPrompt()
	}
}
